zio-profiling/jvm/src/main/scala/zio/profiling/causal/CausalProfiler.scala

```diff
@@ -0,0 +1,313 @@
+    def apply[R, E](zio: ZIO[R, E, Any])(implicit trace: Trace): ZIO[R, E, Result] =
+      ZIO.scoped[R] {
+        supervisor(config).flatMap { prof =>
+          zio
+            .withRuntimeFlags(RuntimeFlags.enable(RuntimeFlag.OpSupervision))
+            .supervised(prof)
+            .fork
+            .zipRight(prof.value)
+        }
+      }
+  }
+
+  def profile[R, E](config: ProfilerConfig): ProfilePartiallyApplied =
+    new ProfilePartiallyApplied(config)
+
+  def supervisor(config: ProfilerConfig)(implicit trace: Trace): ZIO[Scope, Nothing, Supervisor[Result]] = {
+    import config._
+
+    val sleepPrecisionNanos        = sleepPrecision.toNanos()
+    val warmUpPeriodNanos          = warmUpPeriod.toNanos()
+    val coolOffPeriodNanos         = coolOffPeriod.toNanos()
+    val samplingPeriodNanos        = samplingPeriod.toNanos()
+    val minExperimentDurationNanos = minExperimentDuration.toNanos()
+
+    val startTime = java.lang.System.nanoTime()
+
+    val fibers = new ConcurrentHashMap[FiberId, FiberState]()
+
+    @volatile
+    var samplingState: SamplingState = SamplingState.Warmup(startTime + warmUpPeriodNanos)
+
+    @volatile
+    var globalDelay = 0L
+
+    def delayFiber(state: FiberState): Unit =
+      samplingState match {
+        case SamplingState.Done                          =>
+          ()
+        case SamplingState.ExperimentInProgress(_, _, _) =>
+          val delay = globalDelay - state.localDelay.get()
+          if (delay > 0) {
+            val actualSleep = sleepNanos(delay)
+            state.localDelay.addAndGet(actualSleep)
+            ()
+          }
+        case _                                           =>
+          state.localDelay.set(globalDelay)
+      }
+
+    def selectSpeedUp(): Float =
+      if (scala.util.Random.nextInt(zeroSpeedupWeight) == 0) 0L
+      else (scala.util.Random.nextInt(maxConsideredSpeedUp) + 1).toFloat / 100
+
+    // http://andy-malakov.blogspot.com/2010/06/alternative-to-threadsleep.html
+    def sleepNanos(nanoDuration: Long): Long = {
+      val end      = java.lang.System.nanoTime() + nanoDuration
+      var timeLeft = nanoDuration
+
+      while (timeLeft > 0) {
+        if (Thread.interrupted()) {
+          throw new InterruptedException()
+        } else if (timeLeft > sleepPrecisionNanos) {
+          Thread.sleep((timeLeft - sleepPrecisionNanos).nanos.toMillis);
+        }
+        // busy spin
+        timeLeft = end - java.lang.System.nanoTime()
+      }
+      nanoDuration - timeLeft
+    }
+
+    val tracker = new Tracker {
+      def progressPoint(name: String): Unit =
+        samplingState match {
+          case SamplingState.ExperimentInProgress(experiment, _, _) =>
+            experiment.addProgressPointMeasurement(name)
+          case _                                                    =>
+            ()
+        }
+    }
+
+    GlobalTrackerRef.locallyScoped(tracker).zipRight {
```

Comment:
It would be very good to have proper tests for this thing, but it's not clear to me what the best way of approaching them is.

I'll create an issue to add them / look into it